deploy-service/common/src/main/java/com/pinterest/deployservice/alerts/AutoRollbackAction.java

```diff
@@ -0,0 +1,53 @@
+package com.pinterest.deployservice.alerts;
+
+import com.pinterest.deployservice.bean.DeployBean;
+import com.pinterest.deployservice.bean.EnvironBean;
+
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+/**
+ * AutoRollbackAction rollback to the last known good version before action window
+ */
+public class AutoRollbackAction extends AlertAction {
+
+  public final static Logger LOG = LoggerFactory.getLogger(AutoRollbackAction.class);
+
+  //Maximum time looks back. May make it pass in from webhook later
+  private final static int MaxLookbackDays = 30;
+
+  @Override
+  public Object perform(AlertContext context, EnvironBean environ, DeployBean lastDeploy,
+                        int actionWindowInSeconds,
+                        String operator) throws Exception {
+    //It is tricky to do roll back. Here we basically rollback to the last succeed deploy before
+    // the now-window
+    List<DeployBean>
+        candidates =
+        context.getDeployDAO().getAcceptedDeploys(environ.getEnv_id(),
+            new Interval(DateTime.now().minusDays(MaxLookbackDays),
+                DateTime.now().minusSeconds(actionWindowInSeconds)),
+            1);
+    if (candidates.size() > 0) {
+      try {
+        LOG.info("AutoRollback to {}", candidates.get(0).getDeploy_id());
```

Comment:
Could we add env_id in log as well,  easier to debug

Reply:
do we need to filter out deploy with `bad build` from the db result of `getAcceptedDeploys`? Even though `getAcceptedDeploys` only includes acc_status='ACCEPTED' deploys.  I am not sure if there is any possibility that a build is marked as bad, but acc_status is ACCEPTED.

Reply:
This is a good point. I plan to move most of logic from AutoPromoter to DeployHandler. This can ensure we share the same logic on handling deployment selections